File: frontend/src/types/portfolio.ts
/**
 * Portfolio types - UPDATED
 * Currency and Quantity fields removed as requested
 */
import { ApiResponse } from './common';

/**
 * Asset base type
 */
export interface AssetBase {
  ticker: string;
  name?: string;
  weight?: number;
}

/**
 * Asset create request - NO CURRENCY, NO QUANTITY
 */
export interface AssetCreate extends AssetBase {
  purchasePrice?: number;
  purchaseDate?: string;
  currentPrice?: number;
  sector?: string;
  industry?: string;
  assetClass?: string;
  country?: string;
  exchange?: string;
}

/**
 * Asset update request - NO CURRENCY, NO QUANTITY
 */
export interface AssetUpdate {
  name?: string;
  weight?: number;
  purchasePrice?: number;
  purchaseDate?: string;
  currentPrice?: number;
  sector?: string;
  industry?: string;
  assetClass?: string;
  country?: string;
  exchange?: string;
}

/**
 * Asset in portfolio response - QUANTITY CALCULATED AUTOMATICALLY
 */
export interface Asset extends AssetCreate {
  lastUpdated?: string;
  positionValue?: number;
  profitLoss?: number;
  profitLossPct?: number;
  // quantity calculated automatically: (portfolioValue * weight/100) / purchasePrice
  calculatedQuantity?: number;
}

/**
 * Portfolio base type
 */
export interface PortfolioBase {
  name: string;
  description?: string;
  tags?: string[];
}

/**
 * Portfolio create request
 */
export interface PortfolioCreate extends PortfolioBase {
  assets?: AssetCreate[];
  initialValue?: number; // Starting portfolio value for quantity calculations
}

/**
 * Portfolio update request
 */
export interface PortfolioUpdate {
  name?: string;
  description?: string;
  tags?: string[];
  assets?: AssetCreate[];
  resetAssets?: boolean;
  assetsToDelete?: string[];
  initialValue?: number;
}

/**
 * Portfolio response
 */
export interface Portfolio extends PortfolioBase {
  id: string;
  assets: Asset[];
  created: string;
  lastUpdated: string;
  totalValue?: number;
  initialValue?: number;
  performance?: Record<string, number>;
}

/**
 * Portfolio list item
 */
export interface PortfolioListItem {
  id: string;
  name: string;
  description?: string;
  assetCount: number;
  tags: string[];
  lastUpdated: string;
}

/**
 * Text Portfolio Create
 */
export interface TextPortfolioCreate extends PortfolioBase {
  text: string;
  initialValue?: number;
}

/**
 * Update prices response
 */
export interface UpdatePricesResponse {
  portfolioId: string;
  updatedAssets: number;
  timestamp: string;
  priceChanges: Record<string, number>;
}

/**
 * Asset Price Update
 */
export interface AssetPriceUpdate {
  ticker: string;
  currentPrice: number;
  priceDate?: string;
  source?: string;
}

/**
 * Asset Search Result
 */
export interface AssetSearch {
  ticker: string;
  name: string;
  exchange?: string;
  assetType?: string;
  country?: string;
  sector?: string;
  industry?: string;
  // Currency determined automatically by exchange
}

/**
 * Historical Asset Data
 */
export interface AssetHistoricalData {
  ticker: string;
  dates: string[];
  prices: {
    open: number[];
    high: number[];
    low: number[];
    close: number[];
    adjClose: number[];
  };
  volumes?: number[];
  currency: string; // Auto-determined by API
}

/**
 * Asset Performance Data
 */
export interface AssetPerformance {
  ticker: string;
  returns: {
    daily: number;
    weekly: number;
    monthly: number;
    quarterly: number;
    yearly: number;
    ytd: number;
  };
  volatility: number;
  sharpe: number;
  maxDrawdown: number;
  beta?: number;
  currency: string; // Auto-determined by API
}

/**
 * Portfolio Analytics Request
 */
export interface AnalyticsRequest {
  portfolioId: string;
  startDate?: string;
  endDate?: string;
  benchmark?: string;
}

/**
 * Performance Metrics Response
 */
export interface PerformanceMetricsResponse {
  portfolioId: string;
  metrics: {
    totalReturn: number;
    annualizedReturn: number;
    volatility: number;
    sharpeRatio: number;
    maxDrawdown: number;
    calmarRatio: number;
    sortinoRatio: number;
    beta?: number;
    alpha?: number;
    informationRatio?: number;
  };
  periodStart: string;
  periodEnd: string;
  benchmark?: string;
}

/**
 * Risk Metrics Response
 */
export interface RiskMetricsResponse {
  portfolioId: string;
  riskMetrics: {
    valueAtRisk: number;
    conditionalVaR: number;
    standardDeviation: number;
    downside_deviation: number;
    trackingError?: number;
    correlation?: number;
  };
  confidence: number;
  timeHorizon: number;
}

/**
 * Portfolio Comparison Response
 */
export interface PortfolioComparisonResponse {
  portfolio1: string;
  portfolio2: string;
  comparison: {
    performance: {
      returns: Record<string, number>;
      volatility: Record<string, number>;
      sharpe: Record<string, number>;
    };
    allocation: {
      sectorDifferences: Record<string, number>;
      assetClassDifferences: Record<string, number>;
      geographicDifferences: Record<string, number>;
    };
    risk: {
      var: Record<string, number>;
      correlation: number;
      tracking_error: number;
    };
  };
}

/**
 * Asset Form Data Interface (for forms only)
 */
export interface AssetFormData {
  id: string;
  ticker: string;
  name: string;
  weight: number;
  sector?: string;
  assetClass?: string;
  currentPrice?: number;
  purchasePrice?: number;
  purchaseDate?: string;
  industry?: string;
  country?: string;
  exchange?: string;
}

/**
 * Currency Auto-Detection Map
 * Currency is automatically determined by exchange
 */
export const EXCHANGE_CURRENCY_MAP: Record<string, string> = {
  'NASDAQ': 'USD',
  'NYSE': 'USD',
  'LSE': 'GBP',
  'TSE': 'JPY',
  'HKSE': 'HKD',
  'ASX': 'AUD',
  'TSX': 'CAD',
  'FRA': 'EUR',
  'AMS': 'EUR',
  'SWX': 'CHF',
  'BSE': 'INR',
  'NSE': 'INR',
  'SSE': 'CNY',
  'SZSE': 'CNY',
  'KRX': 'KRW',
  'BMV': 'MXN',
  'B3': 'BRL',
};

/**
 * Get currency for exchange
 */
export const getCurrencyForExchange = (exchange: string): string => {
  return EXCHANGE_CURRENCY_MAP[exchange.toUpperCase()] || 'USD';
};

/**
 * Calculate quantity from weight and portfolio value
 */
export const calculateQuantity = (
  weight: number, // in percentage (0-100)
  portfolioValue: number,
  pricePerShare: number
): number => {
  if (pricePerShare <= 0) return 0;
  return Math.floor((portfolioValue * weight / 100) / pricePerShare);
};

/**
 * Calculate position value from quantity and current price
 */
export const calculatePositionValue = (
  quantity: number,
  currentPrice: number
): number => {
  return quantity * currentPrice;
};

/**
 * Calculate weight from position value and total portfolio value
 */
export const calculateWeight = (
  positionValue: number,
  totalPortfolioValue: number
): number => {
  if (totalPortfolioValue <= 0) return 0;
  return (positionValue / totalPortfolioValue) * 100;
};

File: backend/app/schemas/portfolio.py
from pydantic import BaseModel, Field, validator
from typing import List, Dict, Optional, Any
from datetime import datetime
import re


class AssetBase(BaseModel):
    """Base schema for an asset in a portfolio."""
    ticker: str = Field(..., description="Asset ticker symbol")
    weight: Optional[float] = Field(None, description="Asset weight in the portfolio (0-1)")

    @validator('ticker')
    def ticker_must_be_valid(cls, v):
        """Validate that the ticker is a valid stock symbol format."""
        if not re.match(r'^[A-Za-z0-9.\-]+$', v):
            raise ValueError('Ticker must contain only letters, numbers, dots, and hyphens')
        return v.upper()

    @validator('weight')
    def weight_must_be_valid(cls, v):
        """Validate that the weight is between 0 and 1."""
        if v is not None and (v < 0 or v > 1):
            raise ValueError('Weight must be between 0 and 1')
        return v


class AssetCreate(AssetBase):
    """Schema for creating a new asset."""
    name: Optional[str] = Field(None, description="Asset name")
    quantity: Optional[float] = Field(None, description="Asset quantity")
    purchase_price: Optional[float] = Field(None, description="Purchase price")
    purchase_date: Optional[datetime] = Field(None, description="Purchase date")
    current_price: Optional[float] = Field(None, description="Current price")
    sector: Optional[str] = Field(None, description="Sector")
    industry: Optional[str] = Field(None, description="Industry")
    asset_class: Optional[str] = Field(None, description="Asset class")
    currency: Optional[str] = Field("USD", description="Currency")
    country: Optional[str] = Field(None, description="Country")
    exchange: Optional[str] = Field(None, description="Exchange")


class AssetResponse(AssetCreate):
    """Schema for an asset response."""
    price_change_pct: Optional[float] = Field(None, description="Price change percentage")
    price_date: Optional[str] = Field(None, description="Price date")
    position_value: Optional[float] = Field(None, description="Position value")


class PortfolioBase(BaseModel):
    """Base schema for a portfolio."""
    name: str = Field(..., description="Portfolio name")
    description: Optional[str] = Field("", description="Portfolio description")
    tags: Optional[List[str]] = Field([], description="Portfolio tags")


class PortfolioCreate(PortfolioBase):
    """Schema for creating a new portfolio."""
    assets: List[AssetCreate] = Field([], description="Portfolio assets")


class PortfolioUpdate(BaseModel):
    """Schema for updating an existing portfolio."""
    name: Optional[str] = Field(None, description="Portfolio name")
    description: Optional[str] = Field(None, description="Portfolio description")
    tags: Optional[List[str]] = Field(None, description="Portfolio tags")
    assets: Optional[List[AssetCreate]] = Field(None, description="Portfolio assets")


class PortfolioResponse(PortfolioBase):
    """Schema for a portfolio response."""
    id: str = Field(..., description="Portfolio ID")
    assets: List[AssetResponse] = Field(..., description="Portfolio assets")
    created: datetime = Field(..., description="Creation timestamp")
    last_updated: datetime = Field(..., description="Last update timestamp")
    total_value: Optional[float] = Field(None, description="Total portfolio value")
    performance: Optional[Dict[str, float]] = Field(None, description="Portfolio performance metrics")


class PortfolioList(BaseModel):
    """Schema for a portfolio in a list."""
    id: str = Field(..., description="Portfolio ID")
    name: str = Field(..., description="Portfolio name")
    description: Optional[str] = Field("", description="Portfolio description")
    asset_count: int = Field(..., description="Number of assets")
    tags: List[str] = Field([], description="Portfolio tags")
    last_updated: datetime = Field(..., description="Last update timestamp")


class TextPortfolioCreate(BaseModel):
    """Schema for creating a portfolio from text."""
    name: str = Field(..., description="Portfolio name")
    description: Optional[str] = Field("", description="Portfolio description")
    text: str = Field(..., description="Text with tickers and weights")
    tags: Optional[List[str]] = Field([], description="Portfolio tags")


class PortfolioImport(BaseModel):
    """Schema for importing a portfolio from a file."""
    portfolio_name: str = Field(..., description="Portfolio name")
    description: Optional[str] = Field("", description="Portfolio description")
    file_path: str = Field(..., description="Path to the imported file")
    tags: Optional[List[str]] = Field([], description="Portfolio tags")


class PortfolioExport(BaseModel):
    """Schema for exporting a portfolio to a file."""
    portfolio_id: str = Field(..., description="Portfolio ID")
    format: str = Field("csv", description="Export format (csv, excel, json)")
    file_path: Optional[str] = Field(None, description="Path to save the exported file")


class UpdatePricesResponse(BaseModel):
    """Schema for the response after updating portfolio prices."""
    portfolio_id: str = Field(..., description="Portfolio ID")
    updated_assets: int = Field(..., description="Number of assets updated")
    timestamp: datetime = Field(..., description="Update timestamp")
    price_changes: Dict[str, float] = Field(..., description="Price changes by ticker")